src/vfio/container.h

```diff
@@ -19,12 +19,24 @@ struct vfio_group {
 
 #define VFN_MAX_VFIO_GROUPS 64
 
+enum vfio_iommu_usrapi {
+	VFIO_IOMMU_USRAPI_VFIO,
+	VFIO_IOMMU_USRAPI_IOMMUFD
+};
+
 struct vfio_container {
 	int fd;
 	struct vfio_group groups[VFN_MAX_VFIO_GROUPS];
 	struct iommu_state iommu;
+	uint32_t ioas_id;
+	enum vfio_iommu_usrapi iommu_usrapi;
+	int (*config_iommu)(struct vfio_container *vfio);
+	int (*get_dev_fd)(struct vfio_container *vfio, const char *bdf);
+	int (*do_dma_map)(struct vfio_container *vfio, void *vaddr, size_t len, uint64_t iova);
+	int (*do_dma_unmap)(struct vfio_container *vfio, size_t len, uint64_t iova);
```

Comment:
Since we can have the same prototypes for these functions, I'd like to turn this into a build time thing. As I mentioned below, we can determine if iommufd is available (by checking for `linux/iommufd.h`) and default to using that if it is available. Potentially, we can force "legacy vfio" with a meson `-Ddisable-iommufd` option.

Reply:
I also support the compile-time decision, as they both are feature-complete. This is also in line with my experimental work on porting libvfn to macOS.

Food for thought is whether we should refactor to make a more generic "`vfio_container`", and place it in something like `src/iommu/` as neither iommufd nor my port are using vfio.

Reply:
> Since we can have the same prototypes for these functions, I'd like to turn this into a build time thing. As I mentioned above, we can determine if iommufd is available (by checking for `linux/iommufd.h`) and default to using that if it is available. Potentially, we can force "legacy vfio" with a meson `-Ddisable-iommufd` option.
Yes. this is my intention[

Reply:
> I also support the compile-time decision, as they both are feature-complete. This is also in line with my experimental work on porting libvfn to macOS.
Do you have an example of your code. Maybe I can get inspired.

> Food for thought is whether we should refactor to make a more generic "`vfio_container`", and place it in something like `src/iommu/` as neither iommufd nor my port are using vfio.

I think this makes load of sense as it basically follows the vfio uses iommufd API idea described in the iommufd design 
https://lore.kernel.org/kvm/BN9PR11MB5433B1E4AE5B0480369F97178C189@BN9PR11MB5433.namprd11.prod.outlook.com/

This however might be a bit premature as iommufd is still in the works and is not feature complete. I think for now we can put iommufd functionality in libvfns vfio_container struct and take it out if it makes sense once the patches settle in the kernel. (https://github.com/torvalds/linux/blob/master/Documentation/userspace-api/iommufd.rst under "VFIO and IOMMUFD").

Reply:
All these comments have been included. I will hit resolve on this.